### PR TITLE
Refactor benchmark script

### PR DESCRIPTION
**Pull request by bluwy**
## Changes

Update root `pnpm benchmark` script and CI `!bench` command. See https://github.com/bluwy/astro/pull/2 for examples, e.g. https://github.com/bluwy/astro/pull/2#issuecomment-1448419909

Also added a new `/benchmark` folder with READMEs within it that should hopefully help clarify how it works.

The Astro codebase is also tweaked a bit to collect timing/memory logs via `process.env.ASTRO_TIMER_PATH`, and if it exists collect and write the results to the path. This makes passing the results easier between child processes. It's only used for benchmarking.

---

Locally, you can run `pnpm benchmark`, `pnpm benchmark memory`, `pnpm benchmark server-stress`, etc

In PR comments, you can comment `!bench`, `!bench memory`, `!bench server-stress`, etc

## Testing

Manual testing

## Docs

n/a. This doesn't touch on user documentation

### CODE REVIEW

**Comment by bluwy** [comment]
!bench

**Comment by bluwy** [comment]
Looks like I need to fork this to properly test the workflow. Now it's using the main workflow.

**Comment by natemoo-re** [comment]
> Looks like I need to fork this to properly test the workflow. Now it's using the main workflow.

Yep that's correct. I use a personal fork for testing actions/CI.

**Review by matthewp — APPROVED** [review]
Looks great! Excited for this.

File: benchmark/bench/server-stress.js
Comment (ematipico) [inline benchmark/bench/server-stress.js:50]: Would you mind adding a small comment explaining what we're doing here and why?

File: benchmark/ci-helper.js
Comment (ematipico) [inline benchmark/ci-helper.js:1]: Could you add a comment to this file explaining what it is for? It seems to be a particular file used only in CI for doing something that I can't understand 😅

File: packages/astro/src/core/config/timer.ts
Comment (ematipico) [inline packages/astro/src/core/config/timer.ts:21]: Maybe we could use `WeakMap`? It doesn't seem we iterate them

**Review by ematipico — APPROVED** [review]

File: benchmark/make-project/README.md
Comment (sarah11918) [inline benchmark/make-project/README.md:5]:
```suggestion
Each benchmark can specify the default project to run in its `defaultProject` export, but it can be overriden if `--project <project-name>` is passed through the CLI.
```

**Review by sarah11918 — COMMENTED** [review]
Just a heads up that Docs usually removes the review request to Astro Docs Maintainers ourselves! (We get pinged, so we do monitor these requests.)

We are automatically pinged whenever a README file is included in a PR because this is content that someone might... read! And even if internal, we do care about its quality.

File: benchmark/bench/README.md
Comment (bluwy) [reply benchmark/bench/README.md:5]: Ah sorry about that! Noted 👍 Thanks for catching these grammar mistakes!

File: benchmark/bench/server-stress.js
Comment (bluwy) [reply benchmark/bench/server-stress.js:50]: Good idea 👍 It's here mostly to truncate the markdown text in the comments so it's not huge. I'll update this.

File: benchmark/ci-helper.js
Comment (bluwy) [reply benchmark/ci-helper.js:1]: Yup! I've written this mostly because I can't write the same thing in bash 😅 I'll note what it's doing

File: packages/astro/src/core/config/timer.ts
Comment (bluwy) [reply packages/astro/src/core/config/timer.ts:21]: Weak maps only allows objects as keys, we also need to refetch the `OngoingStat` when calling the `end()` method, so we need it to persists indefinitely as there isn't a reference to pass around.

**Comment by bluwy** [comment]
Going to merge this to unblock the memory optimizations soon. Happy to improve further if there's more changes to add. Thanks for the reviews!

**Pull request merged**
